libretroshare/src/serialiser/rstypeserializer.h

```diff
@@ -439,176 +454,222 @@ struct RsTypeSerializer
+				RsGenericSerializer::SerializeContext elCtx(
+				            nullptr, 0, ctx.mFlags, &allocator );
+				serial_process(j, elCtx, el, memberName);
+
+				elCtx.mOk = elCtx.mOk && elCtx.mJson.HasMember(arrKey);
+				if(elCtx.mOk) arr.PushBack(elCtx.mJson[arrKey], allocator);
+				else
+				{
+					ctx.mOk = false;
+					break;
+				}
 			}
+
+			ctx.mJson.AddMember(arrKey, arr, allocator);
 			break;
 		}
-		case RsGenericSerializer::SERIALIZE:
-		{
-			uint32_t n=v.size();
-			serial_process(j,ctx,n,"temporary size");
-			for(typename std::set<T>::iterator it(v.begin());it!=v.end();++it)
-				// the const cast here is a hack to avoid serial_process to
-				// instantiate serialise(const T&)
-				serial_process(j,ctx,const_cast<T&>(*it) ,memberName);
-			break;
-		}
-		case RsGenericSerializer::PRINT:
+		case RsGenericSerializer::FROM_JSON:
 		{
-			if(v.empty())
-				std::cerr << "  Empty set \"" << memberName << "\""
-				          << std::endl;
-			else
-				std::cerr << "  Set of " << v.size() << " elements: \""
-				          << memberName << "\"" << std::endl;
+			using namespace rapidjson;
+
+			bool ok = ctx.mOk || !!(ctx.mFlags & RsSerializationFlags::YIELDING);
+			Document& jDoc(ctx.mJson);
+			Document::AllocatorType& allocator = jDoc.GetAllocator();
+
+			Value arrKey;
+			arrKey.SetString( memberName.c_str(),
+			                  static_cast<SizeType>(memberName.length()) );
+
+			ok = ok && jDoc.IsObject() && jDoc.HasMember(arrKey)
+			        && jDoc[arrKey].IsArray();
+			if(!ok) { ctx.mOk = false; break; }
+
+			for (auto&& arrEl : jDoc[arrKey].GetArray())
+			{
+				Value arrKeyT;
+				arrKeyT.SetString(
+				            memberName.c_str(),
+				            static_cast<SizeType>(memberName.length()) );
+
+				RsGenericSerializer::SerializeContext elCtx(
+				            nullptr, 0, ctx.mFlags, &allocator );
+				elCtx.mJson.AddMember(arrKeyT, arrEl, allocator);
+
+				el_t el;
+				serial_process(j, elCtx, el, memberName);
+				ok = ok && elCtx.mOk;
+				ctx.mOk &= ok;
+				if(ok) member.insert(member.end(), el);
+				else break;
+			}
+
 			break;
 		}
-		case RsGenericSerializer::TO_JSON:
-			RsTypeSerializer_PRIVATE_TO_JSON_ARRAY();
-			break;
-		case RsGenericSerializer::FROM_JSON:
-			RsTypeSerializer_PRIVATE_FROM_JSON_ARRAY(insert);
-			break;
 		default: fatalUnknownSerialJob(j);
 		}
 	}
 
-	/// std::list<T>
+//============================================================================//
+//                               Strings                                      //
+//============================================================================//
+
+	/// Strings
 	template<typename T>
-	static void serial_process( RsGenericSerializer::SerializeJob j,
-	                            RsGenericSerializer::SerializeContext& ctx,
-	                            std::list<T>& v,
-	                            const std::string& memberName )
+	typename std::enable_if<is_string<T>::value>::type
```

Comment:
Which other types than std::string are you targetting here?

Reply:
ATM we have only std::string but the idea is that this code is generic enough to support other kind of strings implementation, think of std::basic_string<char, Our_custom_allocator> which is very similar to std::string except for using a specialized allocator